src/java.base/share/classes/jdk/internal/io/JdkConsoleImpl.java

```diff
@@ -164,37 +151,32 @@ public void run() {
         shutdownHookInstalled = true;
     }
 
-    /**
-     * {@inheritDoc}
-     */
     @Override
     public char[] readPassword() {
         return readPassword("");
     }
 
-    /**
-     * {@inheritDoc}
-     */
     @Override
     public void flush() {
         pw.flush();
     }
 
-    /**
-     * {@inheritDoc}
-     */
     @Override
     public Charset charset() {
-        assert CHARSET != null : "charset() should not return null";
-        return CHARSET;
+        assert charset != null : "charset() should not return null";
+        return charset;
     }
 
-    private final Object readLock;
-    private final Object writeLock;
-    private final Reader reader;
-    private final Writer out;
-    private final PrintWriter pw;
-    private final Formatter formatter;
+    // The singleton
+    private volatile static JdkConsole INSTANCE;
+
+    private Charset charset;
+    private Object readLock;
+    private Object writeLock;
+    private Reader reader;
+    private Writer out;
+    private PrintWriter pw;
+    private Formatter formatter;
```

Comment:
Hello Naoto, I think some of these can be made `final`.

Reply:
Actually, I see that it isn't possible for these to be final because we set these up in the `console(...)` instance method of this class. That's because this `JdkConsoleImpl` implements both the `JdkConsole` interface and the `JdkConsoleProvider` interface.

Reply:
Do you think we could separate out this class into 2 separate implementations? Just like we do in the `jdk.internal.org.jline.JdkConsoleProviderImpl`? That could perhaps make this implementation similar to that other one. So something like:

```
package jdk.internal.io;
public final class JdkConsoleProviderImpl implements JdkConsoleProvider {
	
	private static volatile JdkConsole INSTANCE;

	@Override
    public JdkConsole console(boolean isTTY, Charset charset) {
    	.....
    	// construct and return a (JdkConsoleImpl instance)
    }

    public static synchronized JdkConsole getJdkConsole(Charset cs) {
        return INSTANCE != null ? INSTANCE : new JdkConsoleImpl(cs);
    }

    private static final class JdkConsoleImpl implements JdkConsole {
    	private final Charset charset;
    	private final Object readLock;
	    private final Object writeLock;
	    private final Reader reader;
	    private final Writer out;
	    private final PrintWriter pw;
	    private final Formatter formatter;
	    private char[] rcb;
	    private boolean restoreEcho;
	    private boolean shutdownHookInstalled;


	    private JdkConsoleImpl(Charset cs) {
	    	this.charset = cs;
	    	....
	    }

	    // other methods from JdkConsole

    }


}
```

Reply:
Thanks, Jai. Yes, the reason I dropped those `final`s were they are not final. Anyway I got rid of `JdkConsoleProviderImpl` altogether, as we know this is the last resort impl, so no need to have `ServiceLoader` to instantiate it.